iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+  auto lhsFuncOps = llvm::to_vector<1>(lhsModule.getOps<FuncOp>());
+  auto rhsFuncOps = llvm::to_vector<1>(rhsModule.getOps<FuncOp>());
+  if (lhsFuncOps.size() != rhsFuncOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsFuncOps.size(); ++i) {
+    auto lhsRegion = lhsFuncOps[i].getCallableRegion();
+    auto rhsRegion = rhsFuncOps[i].getCallableRegion();
+    return areRegionsEquivalent(lhsRegion, rhsRegion);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+class DeduplicateExecutablesPass
+    : public PassWrapper<DeduplicateExecutablesPass, OperationPass<ModuleOp>> {
+ public:
+  explicit DeduplicateExecutablesPass() {}
+  DeduplicateExecutablesPass(const DeduplicateExecutablesPass &pass) {}
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto executableOps = llvm::to_vector<8>(moduleOp.getOps<ExecutableOp>());
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    SmallVector<ExecutableOp, 3> duplicateExecutableOps;
+    DenseMap<Attribute, Attribute> entryPointRefReplacements;
+
+    // For each executable, find the first executable which it is equivalent to.
+    for (int i = executableOps.size() - 1; i >= 0; --i) {
+      auto executableOp = executableOps[i];
+
+      for (int j = 0; j < i; ++j) {
+        auto comparisonExecutableOp = executableOps[j];
+
+        if (!areExecutablesEquivalent(executableOp, comparisonExecutableOp)) {
+          continue;
+        }
+
+        // Found an equivalent executable! Record it and move on to the next.
+        duplicateExecutableOps.push_back(executableOp);
+
+        // Record entry point reference replacements.
+        auto dispatchEntryOps = llvm::to_vector<1>(
+            executableOp.getBlock().getOps<DispatchEntryOp>());
+        auto comparisonDispatchEntryOps = llvm::to_vector<1>(
+            comparisonExecutableOp.getBlock().getOps<DispatchEntryOp>());
+        assert(dispatchEntryOps.size() == comparisonDispatchEntryOps.size());
+        for (int k = 0; k < dispatchEntryOps.size(); ++k) {
+          auto oldSymbolRefAttr = builder.getSymbolRefAttr(
+              executableOp.getName(),
+              {builder.getSymbolRefAttr(dispatchEntryOps[k].sym_name())});
+          auto newSymbolRefAttr = builder.getSymbolRefAttr(
+              comparisonExecutableOp.getName(),
+              {builder.getSymbolRefAttr(
+                  comparisonDispatchEntryOps[k].sym_name())});
+          entryPointRefReplacements[oldSymbolRefAttr] = newSymbolRefAttr;
+        }
+
+        break;
+      }
+    }
+
+    totalExecutables = executableOps.size();
+    executablesDeduplicated = duplicateExecutableOps.size();
+    remainingExecutables = totalExecutables - executablesDeduplicated;
+
+    replaceEntryPointUses(moduleOp, entryPointRefReplacements);
+    for (auto executableOp : duplicateExecutableOps) {
+      executableOp.erase();
+    }
+
+    // TODO(scotttodd): rewrite executable indices, filling in gaps?
```

Comment:
it's useful to be able to track them through the deduplication (at least one way) - like dispatch_43 at the end will be dispatch_43 in the beginning - it's not always possible to preserve them like that, but  it'd be nice not to intentionally make them differ :)

Reply:
👍 changed the TODO into a comment explaining that